### PR TITLE
[FIX] Ledger login and redirect improvements + Bluetooth Connect support

### PR DESCRIPTION
**Pull request by razvantomegea**
### Feature
Ledger login improvements:
- avoid multiple instantiations and connections to ledger transport;
- avoid page reload after login redirect if callbackRoute is in the same origin;
- support for Ledger Bluetooth connection.

### Contains breaking changes
[x] No

[] Yes

### Updated CHANGELOG
[x] Yes

### Testing
[x] User testing
[x] Unit tests

### CODE REVIEW

File: src/UI/DappModal/components/DappModal.tsx [hunk 1]
```diff
@@ -32,7 +32,8 @@ export const DappModal = ({
   parentElement,
   config = defaultConfig,
   children,
-  className = 'dapp-modal-dialog-wrapper'
+  className = 'dapp-modal-dialog-wrapper',
+  'data-testid': dataTestId = 'dappModal'
```
Comment (arhtudormorar) [inline src/UI/DappModal/components/DappModal.tsx:36]: use from Enum

File: src/components/ProviderInitializer/hooks/useSetLedgerProvider.ts [hunk 2]
```diff
@@ -36,27 +21,51 @@ export const useSetLedgerProvider = () => {
...
+      provider instanceof HWProvider && provider.isInitialized();
+
+    let hwWalletP = provider as HWProvider;
+
+    if (!isInitialized) {
+      hwWalletP = new HWProvider();
+      isInitialized = await hwWalletP.init();
+    }
+
+    if (!isInitialized || !hwWalletP) {
```
Comment (arhtudormorar) [inline src/components/ProviderInitializer/hooks/useSetLedgerProvider.ts:43]: name condition

File: src/components/ProviderInitializer/hooks/useSetLedgerProvider.ts [hunk 3]
```diff
@@ -36,27 +21,51 @@ export const useSetLedgerProvider = () => {
...
+      return;
+    }
+
+    if (ledgerLogin?.index != null) {
```
Comment (arhtudormorar) [inline src/components/ProviderInitializer/hooks/useSetLedgerProvider.ts:47]: name condition

File: src/hooks/login/useLedgerLogin.ts [hunk 4]
```diff
@@ -61,6 +60,7 @@ export function useLedgerLogin({
   onLoginRedirect
 }: UseLedgerLoginPropsType): LedgerLoginHookReturnType {
   const ledgerAccount = useSelector(ledgerAccountSelector);
+  const hwProvider = getAccountProvider() as HWProvider;
```
Comment (arhtudormorar) [inline src/hooks/login/useLedgerLogin.ts:63]: why cast?

File: src/hooks/login/useLedgerLogin.ts [hunk 5]
```diff
@@ -120,20 +115,60 @@ export function useLedgerLogin({
...
+    }
+
+    return false;
   };
 
-  async function loginUser(hwWalletProvider: HWProvider) {
-    if (selectedAddress == null) {
-      return false;
+  const initHWProvider = async () => {
+    if (await isHWProviderInitialized()) {
```
Comment (arhtudormorar) [inline src/hooks/login/useLedgerLogin.ts:144]: extract to isInitialized = await..

File: src/hooks/login/useLedgerLogin.ts [hunk 6]
```diff
@@ -120,20 +115,60 @@ export function useLedgerLogin({
...
+        setAccountProvider(hwWalletP);
+      }
+
+      setError('');
+      setIsLoading(false);
+    } catch (e) {
+      onLoginFailed(e);
+    }
+
+    return;
```
Comment (arhtudormorar) [inline src/hooks/login/useLedgerLogin.ts:164]: return?

File: src/hooks/login/useLedgerLogin.ts [hunk 7]
```diff
@@ -195,108 +235,111 @@ export function useLedgerLogin({
...
       setIsLoading(false);
     }
+
     setShowAddressList(false);
+
     return true;
   }
 
   async function fetchAccounts() {
+    if (!(await isHWProviderInitialized())) {
```
Comment (arhtudormorar) [inline src/hooks/login/useLedgerLogin.ts:252]: extract to constant

File: src/hooks/login/useLedgerLogin.ts [hunk 8]
```diff
@@ -195,108 +235,111 @@ export function useLedgerLogin({
...
       setIsLoading(false);
     } catch (err) {
-      const { errorMessage, defaultErrorMessage } = getLedgerErrorCodes(err);
-      setError(errorMessage ?? defaultErrorMessage);
-      console.error('error', err);
-      setIsLoading(false);
+      onLoginFailed(err);
     }
+
+    return;
```
Comment (arhtudormorar) [inline src/hooks/login/useLedgerLogin.ts:277]: no need for return

File: src/hooks/login/useLedgerLogin.ts [hunk 9]
```diff
@@ -195,108 +235,111 @@ export function useLedgerLogin({
...
+    initHWProvider();
+  }, [hwProvider]);
+
   useEffect(() => {
     fetchAccounts();
-  }, [startIndex]);
+  }, [startIndex, showAddressList, hwProvider]);
+
+  useEffect(() => {
+    if (accounts?.length && !showAddressList) {
```
Comment (arhtudormorar) [inline src/hooks/login/useLedgerLogin.ts:339]: extract to constant. 
better not use accounts?.length directly, but with > 0

File: src/utils/redirect.ts [hunk 10]
```diff
@@ -11,7 +11,11 @@ export const safeRedirect = (url: string, timeout = 0) => {
         return;
       }
 
-      window.location.assign(url);
+      if (url.startsWith('/')) {
+        return window.history.pushState('', '', url);
```
Comment (arhtudormorar) [inline src/utils/redirect.ts:15]: add comment on why this is necessary

File: src/hooks/login/useLedgerLogin.ts (hunk 4 above)
Comment (razvantomegea) [reply src/hooks/login/useLedgerLogin.ts:63]: `getAccountProvider()` has multiple types and dynamic cast is required

File: src/components/ProviderInitializer/hooks/useSetLedgerProvider.ts [hunk 11]
```diff
@@ -36,27 +21,59 @@ export const useSetLedgerProvider = () => {
...
   const ledgerLogin = useSelector(ledgerLoginSelector);
+  const provider = getAccountProvider();
 
   const [ledgerData, setLedgerData] =
     useState<{
       version: string;
       dataEnabled: boolean;
     }>();
 
+  const initHWProvider = async () => {
```
Comment (CiprianDraghici) [inline src/components/ProviderInitializer/hooks/useSetLedgerProvider.ts:32]: `init` and `setAddressIndex` methods could throw any exception ? If yes, I think a try-catch block would be very useful

Comment (razvantomegea) [reply src/components/ProviderInitializer/hooks/useSetLedgerProvider.ts:32]: Resolved

File: src/utils/redirect.ts (hunk 10 above)
Comment (razvantomegea) [reply src/utils/redirect.ts:19]: Resolved

File: src/hooks/login/useLedgerLogin.ts (hunk 9 above)
Comment (razvantomegea) [reply src/hooks/login/useLedgerLogin.ts]: Resolved

File: src/hooks/login/useLedgerLogin.ts (hunk 8 above)
Comment (razvantomegea) [reply src/hooks/login/useLedgerLogin.ts]: Resolved

File: src/hooks/login/useLedgerLogin.ts (hunk 7 above)
Comment (razvantomegea) [reply src/hooks/login/useLedgerLogin.ts]: Resolved

File: src/hooks/login/useLedgerLogin.ts (hunk 6 above)
Comment (razvantomegea) [reply src/hooks/login/useLedgerLogin.ts]: Resolved

File: src/hooks/login/useLedgerLogin.ts (hunk 5 above)
Comment (razvantomegea) [reply src/hooks/login/useLedgerLogin.ts]: Resolved

File: src/hooks/login/useLedgerLogin.ts (hunk 4 above)
Comment (razvantomegea) [reply src/hooks/login/useLedgerLogin.ts:63]: Resolved

File: src/components/ProviderInitializer/hooks/useSetLedgerProvider.ts (hunk 3 above)
Comment (razvantomegea) [reply src/components/ProviderInitializer/hooks/useSetLedgerProvider.ts]: Resolved

File: src/UI/DappModal/components/DappModal.tsx (hunk 1 above)
Comment (razvantomegea) [reply src/UI/DappModal/components/DappModal.tsx]: Resolved

**Review by tdumitrache — APPROVED** [review]

**Review by CiprianDraghici — APPROVED** [review]

**Pull request merged**